### PR TITLE
Working example

### PR DESCRIPTION
**Pull request by bouweandela**
To review, follow the instructions in [README.md](https://github.com/nlesc-recruit/usage-example-locally-installed/blob/working-example/README.md) and check that everything works as expected.

### CODE REVIEW

**Comment by svlugt** [comment]
I have followed the build instructions from the above mentioned README on a DAS6 GPU node at our site with `cmake version 3.20.2`, `gcc/9.4.0 `and `cuda/11.6.2`. Build of both the cudawrappers library as well as the locally installed example from this repository seem to pass but upon running the example I get an error: cu::Error: the provided PTX was compiled with an unsupported toolchain.

**Comment by bouweandela** [comment]
Thanks for trying this out @svlugt!

I'm no expert on this, but I think that the [cuda driver may be too old for the cuda toolkit that you're running](https://forums.developer.nvidia.com/t/provided-ptx-was-compiled-with-an-unsupported-toolchain-error-using-cub/168292). When I run `nvidia-smi` on the DAS6 astron cluster, it reports a driver version of 495.29.05. From the cuda toolkit release notes, it looks like this driver version is only supported up to [cuda toolkit 11.5](https://docs.nvidia.com/cuda/archive/11.5.2/cuda-toolkit-release-notes/index.html). 11.6.2 requires at least [510.47.03](https://docs.nvidia.com/cuda/archive/11.6.2/cuda-toolkit-release-notes/index.html)

**Comment by svlugt** [comment]
@bouweandela there is indeed a mis match in my CUDA version and the driver, sorry for not picking this up!
I have re-build with an older version of CUDA, now I get another error though: `cu::Error: named symbol not found`. This still might be an issue from my side I'll look in to this when I have a bit more time and get back to you.

**Comment by bouweandela** [comment]
Could you post the commands that you're running here? Then I can try to reproduce the problem.

**Comment by svlugt** [comment]
@bouweandela we have installed the new CUDA drivers on DAS6 and I am now using CUDA 11.6.2 with both the cudawrappers library and the example.
I have followed the build instructions from the cudawarappers repository and this repository.

I get different results depending on the location from where I'm running the vector_add example.
Running from my home directory:
```
~ $ vector_add_example
cu::Error: named symbol not found
```
Running from within the example directory:
```
~/usage-example-locally-installed $ vector_add_example
hurray! 2
```

So the example application works, but not in all situations.

The problem seems to be that the .cu file is not found when the example is run from another directory. Similarly to what is described in issue [116 ](https://github.com/nlesc-recruit/cudawrappers/issues/116).

As this is quite a common situation I do think that it's good to add to the example how to deal with this.

A solution might be to install the kernel sources along with the binary in the install directory and parse the kernel path with a cmake runtime variable to the application.

**Comment by john-romein** [comment]
Also make sure that the CUDA source can be found when running through nv-nsight-cu.  Unfortunately, the profiler changes the current directory under some circumstances, so that the source file cannot be found while it would have been found when not running via nv-nsight-cu.  This was exactly the reason why I embed the CUDA code in an object file so that it is inside the executable and can always be found, but embedding CUDA code in object files significantly complicates the building processes, and may cause its own problems.

**Comment by bouweandela** [comment]
Thanks for trying this out @svlugt and @john-romein! I included the cuda code as a string now. Does this look like a good solution to you?

I will also have a look if I can improve the error message `cu::Error: named symbol not found` in the nvrtc library, something like `file '/path/to/file.cu' not found` would be much more clear.

**Comment by bouweandela** [comment]
A pull request with an improved error message is available here: https://github.com/nlesc-recruit/cudawrappers/pull/159

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+find_package(cudawrappers 0.5.0 REQUIRED cu nvrtc)
+
+add_executable(vector_add_example src/vector_add.cpp)
+
+target_link_libraries(vector_add_example PRIVATE cudawrappers::cu
+                                                 cudawrappers::nvrtc)
+
+# Make it possible to #include cuda source code
+function(include_cuda_code target input_file)
```
Comment (matmanc) [inline CMakeLists.txt:27]: Would it be an idea to add this function in the cudawrappers CMakefile? So that also other examples could benefit?

**Comment by csbnw** [comment]
A follow-up MR is coming soon, so I will merge this to prepare for it.

**Pull request merged**